product_listing.go

```diff
@@ -0,0 +1,147 @@
+package goshopify
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const productListingBasePath = "product_listings"
+const productsListingResourceName = "product_listings"
+
+// linkRegex is used to extract pagination links from product search results.
+//var linkRegex = regexp.MustCompile(`^ *<([^>]+)>; rel="(previous|next)" *$`)
+
+// ProductListingService is an interface for interfacing with the product listing endpoints
+// of the Shopify API.
+// See: https://help.shopify.com/api/reference/product_listings
+type ProductListingService interface {
+	List(interface{}) ([]ProductListing, error)
+	ListWithPagination(interface{}) ([]ProductListing, *Pagination, error)
+	Count(interface{}) (int, error)
+	Get(int64, interface{}) (*ProductListing, error)
+	GetProductIDs(interface{}) ([]int64, error)
+	//GetProductIDsWithPagination????() ([]int64, *Pagination, error)
+	Publish(int64) (*ProductListing, error)
+	Delete(int64) error
+}
+
+// ProductListingServiceOp handles communication with the product related methods of
+// the Shopify API.
+type ProductListingServiceOp struct {
+	client *Client
+}
+
+// ProductListing represents a Shopify product published to your sales channel app
+type ProductListing struct {
+	ID          int64           `json:"product_id,omitempty"`
+	Title       string          `json:"title,omitempty"`
+	BodyHTML    string          `json:"body_html,omitempty"`
+	Vendor      string          `json:"vendor,omitempty"`
+	ProductType string          `json:"product_type,omitempty"`
+	Handle      string          `json:"handle,omitempty"`
+	CreatedAt   *time.Time      `json:"created_at,omitempty"`
+	UpdatedAt   *time.Time      `json:"updated_at,omitempty"`
+	PublishedAt *time.Time      `json:"published_at,omitempty"`
+	Tags        string          `json:"tags,omitempty"`
+	Options     []ProductOption `json:"options,omitempty"`
+	Variants    []Variant       `json:"variants,omitempty"`
+	Images      []Image         `json:"images,omitempty"`
+}
+
+// Represents the result from the product_listings/X.json endpoint
+type ProductListingResource struct {
+	ProductListing *ProductListing `json:"product_listing"`
+}
+
+// Represents the result from the product_listings.json endpoint
+type ProductsListingsResource struct {
+	ProductListings []ProductListing `json:"product_listings"`
+}
+
+// Represents the result from the product_listings/product_ids.json endpoint
+type ProductListingIDsResource struct {
+	ProductIDs []int64 `json:"product_ids"`
+}
+
+// Resource which create product_listing endpoint expects in request body
+// e.g.
+// PUT /admin/api/2020-07/product_listings/921728736.json
+// {
+//   "product_listing": {
+//     "product_id": 921728736
+//   }
+// }
+type ProductListingPublishResource struct {
+	ProductListing struct {
+		ProductID int64 `json:"product_id"`
+	} `json:"product_listing"`
+}
```

Comment:
Because the ProductID int64 type is used in multiple places I would suggest creating a type for it as follows:

type ProductID struct {
	ProductID int64 `json:"product_id"`
}

type ProductListingIDsResource struct {
	ProductIDs []ProductID `json:"product_ids"`
}

type ProductListingPublishResource struct {
	ProductListing struct {
		ProductID
	} `json:"product_listing"`
}

Reply:
The issue with this approach is the following endpoint which returns the ids as thought they are an array of ints rather than objects. What do you think should I modify the other structs but leave ProductListingIDsResource as is?

`GET /admin/api/2020-10/product_listings/product_ids.json

HTTP/1.1 200 OK
{
  "product_ids": [
    921728736,
    632910392
  ]
}`

Reply:
I do not think a new type is needed. 
It is possible to do something like `type PlatformID int`, but I do not see any benefits of doing this